lib/SILOptimizer/Transforms/AllocBoxToStack.cpp

```diff
@@ -275,37 +281,74 @@ static bool partialApplyEscapes(SILValue V, bool examineApply) {
 }
 
 static SILInstruction *findUnexpectedBoxUse(SILValue Box,
-                                            bool examinePartialApply,
                                             bool inAppliedFunction,
-                                            SmallVectorImpl<Operand *> &);
+                                            SmallVectorImpl<Operand *> &,
+                                            SmallPtrSetImpl<SILFunction *> &,
+                                            unsigned CurrentRecurDepth);
 
-/// checkPartialApplyBody - Check the body of a partial apply to see
+/// checkLocalApplyBody - Check the body of an apply's callee to see
 /// if the box pointer argument passed to it has uses that would
 /// disqualify it from being promoted to a stack location.  Return
-/// true if this partial apply will not block our promoting the box.
-static bool checkPartialApplyBody(Operand *O) {
+/// true if this apply will not block our promoting the box.
+static bool checkLocalApplyBody(Operand *O,
+                                SmallVectorImpl<Operand *> &PromotedOperands,
+                                SmallPtrSetImpl<SILFunction*> &VisitedCallees,
+                                unsigned CurrentRecurDepth) {
   SILFunction *F = ApplySite(O->getUser()).getReferencedFunctionOrNull();
   // If we cannot examine the function body, assume the worst.
   if (!F || F->empty())
     return false;
 
-  // We don't actually use these because we're not recursively
-  // rewriting the partial applies we find.
-  SmallVector<Operand *, 1> PromotedOperands;
+  auto iter = VisitedCallees.insert(F);
+  if (!iter.second)
+    return false;
+
   auto calleeArg = F->getArgument(ApplySite(O->getUser()).getCalleeArgIndex(*O));
-  return !findUnexpectedBoxUse(calleeArg, /* examinePartialApply = */ false,
-                               /* inAppliedFunction = */ true,
-                               PromotedOperands);
+  auto res =
+      !findUnexpectedBoxUse(calleeArg,
+                            /* inAppliedFunction = */ true, PromotedOperands,
+                            VisitedCallees, CurrentRecurDepth + 1);
+  return res;
+}
+
+// Returns true if a callee is eligible to be cloned and rewritten for
+// AllocBoxToStack opt. We don't want to increase code size, so this is
+// restricted only for private local functions currently.
+static bool isEligibleApply(ApplySite Apply) {
+  auto callee = Apply.getReferencedFunctionOrNull();
+  if (!callee) {
+    return false;
+  }
+
+  // Callee should be optimizable.
+  if (!callee->shouldOptimize())
+    return false;
+
+  // External function definitions.
+  if (!callee->isDefinition())
+    return false;
+
+  // Do not optimize always_inlinable functions.
+  if (callee->getInlineStrategy() == Inline_t::AlwaysInline)
+    return false;
+
+  if (callee->getLinkage() != SILLinkage::Private ||
+      !callee->getDeclContext() ||
+      !callee->getDeclContext()->isLocalContext()) {
+    return false;
```

Comment:
callee->getDeclContext()->isLocalContext() is used to check if a callee is a local function. callee->getDeclContext() will not be available for a SIL test. @gottesmm I didn't add a SIL test for this reason.

Reply:
We should add a way of doing this in SIL. Maybe add a function attribute? And make a helper that checks first for the function attribute and then the decl context? Or maybe we assert that they are the same?

Reply:
Looked at this further and getDeclContext()->isLocalContext() covers a much larger case. 
The fact that the linkage is private and a function has a box argument should be sufficient. Looks like only local functions with captures of outer function locals have arguments of SILBoxType. I added a SIL test.